pkg/util/workqueue/prometheus/prometheus.go

```diff
@@ -25,13 +25,107 @@ import (
 // Package prometheus sets the workqueue DefaultMetricsFactory to produce
 // prometheus metrics. To use this package, you just have to import it.
 
+// Metrics subsystem and keys used by the workqueue.
+const (
+	WorkQueueSubsystem         = "workqueue"
+	DepthKey                   = "depth"
+	AddsKey                    = "adds_total"
+	QueueLatencyKey            = "queue_latency_seconds"
+	WorkDurationKey            = "work_duration_seconds"
+	UnfinishedWorkKey          = "unfinished_work_seconds"
+	LongestRunningProcessorKey = "longest_running_processor_seconds"
+	RetriesKey                 = "retries_total"
+)
+
 func init() {
 	workqueue.SetProvider(prometheusMetricsProvider{})
 }
 
 type prometheusMetricsProvider struct{}
 
 func (prometheusMetricsProvider) NewDepthMetric(name string) workqueue.GaugeMetric {
+	depth := prometheus.NewGauge(prometheus.GaugeOpts{
+		Subsystem:   WorkQueueSubsystem,
+		Name:        DepthKey,
+		Help:        "Current depth of workqueue",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+	prometheus.Register(depth)
+	return depth
+}
+
+func (prometheusMetricsProvider) NewAddsMetric(name string) workqueue.CounterMetric {
+	adds := prometheus.NewCounter(prometheus.CounterOpts{
+		Subsystem:   WorkQueueSubsystem,
+		Name:        AddsKey,
+		Help:        "Total number of adds handled by workqueue",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+	prometheus.Register(adds)
+	return adds
+}
+
+func (prometheusMetricsProvider) NewLatencyMetric(name string) workqueue.SummaryMetric {
```

Comment:
We should stop using Summary metrics, please use Histogram instead.

Summary metrics can't be aggregated.

Reply:
@loburm what buckets do you prefer or just ignore it now?

Reply:
I'm not familiar with those queues here. I remember that default is for example almost useless for kube-apiserver requests latency, because most of samples are going to first few buckets and wasn't giving enough information for measuring it.

Usually I prefer near 20 reasonable buckets. Let's ask for advice from someone sig-instrumentation. 

@DirectXMan12 @brancz

Reply:
As this is on internal queues, the latencies should be rather small, I'd suggest something along the lines of:

```
prometheus.ExponentialBuckets(10e-9, 10, 10)
```

That gives us exponential buckets from 1 nanosecond to 10 seconds.

Reply:
Not sure that the best approach. I would check current values from few kube-apiservers and select range base on it.

Reply:
if there are histograms for queues in the apiserver, then yes we should be consistent, latency histograms for api requests are very different though from queues, they should be substantially faster

Reply:
@loburm have the conclusion about the buckets ?

Reply:
Do you have any data about current distribution of those samples? But if you are happy with buckets: 
1ns - 10ns
10ns - 100ns...
1s - 10s
Then you can proposal about.

Reply:
After checking current samples, I agree with your proposal.
Code fixed. PTAL @loburm @brancz